ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -502,56 +567,124 @@ initVR cfg = \knownValid (tip, st) lastTicketNo -> ValidationResult {
+      , vrAfter
+      , vrInvalid
+      } = vr
+
+    TxTicket tx tn sn = txTicket
 
 -- | Extend 'ValidationResult' with a new transaction (one which we have not
 -- previously validated) that may or may not be valid in this ledger state.
 extendVRNew :: ApplyTx blk
             => LedgerConfig blk
+            -> SlotNo
+            -- ^ The current 'SlotNo'
             -> GenTx blk
             -> ValidationResult blk
             -> ValidationResult blk
-extendVRNew cfg tx
-         vr@ValidationResult { vrValid
-                             , vrAfter
-                             , vrInvalid
-                             , vrLastTicketNo
-                             , vrNewValid
-                             } =
+extendVRNew cfg currSlot tx vr =
     let nextTicketNo = succ vrLastTicketNo
     in  case runExcept (applyTx cfg tx vrAfter) of
       Left err  -> vr { vrInvalid      = (tx, err) : vrInvalid
                       }
-      Right st' -> vr { vrValid        = vrValid :> TxTicket tx nextTicketNo
+      Right st' -> vr { vrValid        = vrValid :> TxTicket tx nextTicketNo currSlot
                       , vrNewValid     = tx : vrNewValid
                       , vrAfter        = st'
                       , vrLastTicketNo = nextTicketNo
                       }
+  where
+    ValidationResult
+      { vrValid
+      , vrAfter
+      , vrInvalid
+      , vrLastTicketNo
+      , vrNewValid
+      } = vr
 
 -- | Validate internal state
 validateIS :: forall m blk. (IOLike m, ApplyTx blk)
            => MempoolEnv m blk -> STM m (ValidationResult blk)
-validateIS MempoolEnv{mpEnvLedger, mpEnvLedgerCfg, mpEnvStateVar} =
+validateIS mpEnv =
     go <$> getCurrentLedgerState mpEnvLedger <*> readTVar mpEnvStateVar
   where
+    MempoolEnv
+      { mpEnvLedger
+      , mpEnvLedgerCfg
+      , mpEnvStateVar
+      } = mpEnv
+
     go :: LedgerState      blk
        -> InternalState    blk
        -> ValidationResult blk
     go st IS{isTxs, isTip, isLastTicketNo}
         | tip == isTip
         = initVR mpEnvLedgerCfg isTxs (tip, st) isLastTicketNo
         | otherwise
-        = repeatedly (extendVRPrevApplied mpEnvLedgerCfg) (fromTxSeq isTxs)
-        $ initVR mpEnvLedgerCfg TxSeq.Empty (tip, st) isLastTicketNo
+        = repeatedly
+          (extendVRPrevApplied mpEnvLedgerCfg)
+          (txTickets isTxs)
+          $ initVR mpEnvLedgerCfg TxSeq.Empty (tip, st) isLastTicketNo
       where
         tip = Block.pointHash $ ledgerTipPoint st
+
+------------------------------------------------------------------------------
+-- Mempool Transaction Expiry
+------------------------------------------------------------------------------
+
+-- | The number of slots that must pass before a transaction should be
+-- considered expired.
+data ExpiryThreshold
+  = NoExpiryThreshold
+  -- ^ Transactions in the mempool do not expire.
+  -- This is useful for tests in which we don't care about transaction expiry.
+  | ExpiryThreshold !Word64
```

Comment:
Yeah, `ExpiryThreshold 0` probably doesn't make much sense. It's only relevant if you manage to add a transaction to the mempool just before the expiry logic is run in the `onSlotChange` background thread. 

For example, the `ThresholdExpiry` is `0` and the blockchain time ticks to `SlotNo x`. A user manages to have a transaction submitted and accepted at `SlotNo x` just before the `onSlotChange` expiry logic is executed for `SlotNo x`. When that expiry logic executes, the same transaction is instantly expired at `SlotNo x` (remember that transactions expire if `txAcceptedSlotNo <= currentSlotNo - expiryThreshold`).

This seems to be the only scenario in which instant expiry is relevant. Otherwise, it'd act exactly the same as `ExpiryThreshold 1`.

I think I'll add a smart constructor and disallow `ExpiryThreshold 0`.

Reply:
I also noticed that we have a `NumSlots` type so I'll utilize that instead of `Word64`.